pkgs/top-level/metrics.nix: tidy, speed up, and document what's happening in this file

philiptaron opened the pull request:
I took a sick day detour into Nix performance statistics that are tracked on Hydra.

Best reviewed commit by commit.

I validated this was correct by taking a baseline reading before changing anything, then diffing with the end result after all these commits.

Before: 16m41s
After: ~6m37s~ ~5m20s~ 1m59s

That's greater than 8x speedup!

To test it out, run:

```
nix-build pkgs/top-level/release.nix -A metrics
```

- f87776bd7c3cc67de6155ed8c9fd32ebda4bfdb0 Changed to use `stdenvNoCC.mkDerivation` instead of `runCommand`.
- c3f0ef37ebf21bfea5c17a7ef7d53afc39f711bc Added metadata explaining the purpose of `metrics.nix`.
- a34bbb6b9cd5f511fb5d5cff7d2ceb12d927cffb Tidied up variable usage and improved quoting; allowed for unbroken lines.
- 29db1d32be1964eefd223a974f733f261f32ec55 Utilized `__structuredAttrs` in the code.
- b2331161b3aeb84a9a0be008cb1547cb3d34cb56 Removed aggressive variants as they had become identical to non-aggressive variants since Jan 13, 2019.
- 94f1da567e4717c81fc28195a6eb301e720303b6 Configured the `time` command to output JSON and named the JSON files with the `.json` extension.
- b9519e2c0af72fe953c48cfa8e941d558219cfa2 Logged statistics using `jq` with a header included.
- e35ab56228ac9027e54c3eae9267fab48aaccd95 Implemented the `eval-system` parameter for targeting a specific system.
- 146c6b9aadc712723370be6ecee8d28fc8f60331 Switched from XML to JSON for output, ensuring the testing remains the same.
- dfd90bf4c07b6eb3d101190c2ad2369c92ea52d5 Reduced the number of calls to `nix-env -qa` to just once, saving the output for later use with `jq`.
- 251dff3a3e2a057cb7c3f8e46e15e1d95211c012 Added functionality to save raw files and compress the output.
- 4973d53405faece1e5394cc9616fca610641d117 Updated to use the latest Nix version, aligning with CI practices.
- 8ac12754cecbfffadeb97178a32584f9342df04b Disabled aliases during evaluation to prevent numerous warnings, in line with `ci/eval`.
- fda7cefcdbd5e50052dfd3bf9b97ae5cc32d8cf9 Passed `--no-gc-warning` to suppress garbage collection warning messages.
- 32c68bb46e8d51f5833830750af8bbb078299949 Reorganized the script to use the configure, build, and install phases following code review feedback.
- 79feab6cff53917a2fd40be55547faf68ea0eee5 Used the `-print0` option in `find` to handle file names with spaces efficiently.

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

philiptaron committed:
pkgs/top-level/metrics.nix: use stdenvNoCC.mkDerivation instead of runCommand

philiptaron committed:
pkgs/top-level/metrics.nix: add meta explaining this file

philiptaron committed:
pkgs/top-level/metrics.nix: tidy up variables and quoting

We also allow the lines to be unbroken.

philiptaron committed:
pkgs/top-level/metrics.nix: use __structuredAttrs

philiptaron committed:
pkgs/top-level/metrics.nix: remove the aggressive variants

They've been the same as the non-aggressive variants since Jan 13, 2019 (61b8c03a78caa)

philiptaron committed:
pkgs/top-level/metrics.nix: convince the time command to output JSON

Also, name the JSON files with the .json extension and the name of the run being produced.

philiptaron committed:
pkgs/top-level/metrics.nix: log out the statistics using jq with a header

philiptaron committed:
pkgs/top-level/metrics.nix: use the eval-system parameter to target a specific system

philiptaron committed:
pkgs/top-level/metrics.nix: just run nix-env -qa once

If we save the output, we can use it later with a little `jq`.

philiptaron committed:
pkgs/top-level/metrics.nix: save the raw files, compressing the output

philiptaron committed:
pkgs/top-level/metrics.nix: use json, not XML

It's testing the same thing, and it's useful to have the JSON output.

philiptaron committed:
pkgs/top-level/metrics.nix: use the latest Nix version, just like CI

philiptaron committed:
pkgs/top-level/metrics.nix: turn off aliases in the evaluation

This mirrors `ci/eval` and also prevents many warnings from being logged.

philiptaron committed:
pkgs/top-level/metrics.nix: pass --no-gc-warning so that no spew about --add-root shows up

wolfgangwalther reviewed (commented):
Great stuff!

Just a few comments about commit structure, essentially.

philiptaron committed:
pkgs/top-level/metrics.nix: use the configure, build, and install phases

Per CR feedback, this is better than unpack and install with build jammed into install.

philiptaron committed:
pkgs/top-level/metrics.nix: use -print0 to insulate us from spaces in find

philiptaron reviewed (commented):
Thanks for taking a look. I've pushed a new version that fixes the places I missed a clean commit separation.

It made me realize that I don't have a great workflow for commit-stack (what I call patch stack) editing in Git... it's mostly `git rebase -i` and then some hand-editing. Do you use something better?

wolfgangwalther reviewed (approved):
Diff LGTM, did not test.

Awesome work!

philiptaron commented:
I did mess up the counting of broken derivations. Fix here.

- https://github.com/NixOS/nixpkgs/pull/444082
